core/dbt/include/global_project/macros/materializations/helpers.sql

```diff
@@ -72,3 +72,96 @@
   {% endif %}
   {% do return(config_full_refresh) %}
 {% endmacro %}
+
+
+{% macro incremental_validate_on_schema_change(on_schema_change, default_value='ignore') %}
+   
+   {% if on_schema_change not in ['sync', 'append', 'fail', 'ignore'] %}
+     {{ return(default_value) }}
+
+   {% else %}
+     {{ return(on_schema_change) }}
+   
+   {% endif %}
+
+{% endmacro %}
+
+{% macro get_column_names(columns) %}
+  {% set result = [] %}
+  
+  {% for col in columns %}
+    {{ result.append(col.column) }}
+  {% endfor %}
+  
+  {{ return(result) }}
+{% endmacro %}
+
+{% macro diff_columns(source_columns, target_columns) %}
+  {% set result = [] %}
+  {% set source_names = get_column_names(source_columns) %}
+  {% set target_names = get_column_names(target_columns) %}
+   
+   {# check whether the name attribute exists in the target, but dont worry about data type differences #}
```

Comment:
What happens if there _is_ a data type difference? It's possible that the model will fail its incremental build, though it's also possible that the database will neatly handle by coercing the new data into the old data type.

My immediate inclination is to be strict here: If the data types are incompatible, we should consider that a schema change. Depending on the user-supplied value of `on_schema_change`, that could mean raising a compilation error, or attempting to handle within `sync_columns` (i.e. via [`adapter.alter_column_type`](https://github.com/fishtown-analytics/dbt/blob/8c55e744b8e86b282dcea75e5efcf7c15a7142b9/core/dbt/include/global_project/macros/adapters/common.sql#L123)).

(We also have some logic built into the column object that can help with ensuring that numeric / varchar sizes don't get detected as changes. The incremental materialization [already handles](https://github.com/fishtown-analytics/dbt/blob/8c55e744b8e86b282dcea75e5efcf7c15a7142b9/core/dbt/include/global_project/macros/materializations/incremental/incremental.sql#L30-L32) the case where a string column is wider in the new data than in the existing table.)

Reply:
@jtcohen6 Yeah haha this was the piece I was most on-the-fence about in the PR. The scenario you described where the data type changes in the source in such a way that it's incompatible with the target makes sense and is a vulnerability of the current approach, so I agree we should handle that explicitly in the code for `sync_all_columns`. At the very least this should trigger a message to the user. It feels like dropping and recreating the target column with the new type from the source may be a bridge too far, because I *don't* think we should encourage users to mindlessly change data types and expect the system to "handle it". Agree?

Reply:
Depending on the database, it may support altering column types in place (without needing to drop the column + recreate it). Of course, Redshift only supports that for varchar resizing, but other databases are a little cleverer (e.g. [Postgres](https://www.postgresql.org/docs/9.3/ddl-alter.html#AEN2819), [Snowflake](https://docs.snowflake.com/en/sql-reference/sql/alter-table-column.html)) and, in the event of an unsupported type switch, they would throw a more specific error message.

The choice here feels like, Should we try to implicit alter the type (which may or may not return an error), or just raise an error? I'd be happy if the v1 of this cannot account for type changes and always raises an explicit exception.

Reply:
I'd be happy with punting on data type differences for the v1 of this, documenting it as a known limitation, opening a ticket to come back to it